0013-secure-http-api.adoc

```diff
@@ -52,34 +52,98 @@ The industry standard for encrypting traffic over a network is Transport Layer S
+
+=== Self signed certificate
+
+Use case: connecting to `comit-i` with the machine running cnd.
+Create a self signed certificate.
+
+.pros
+* Free
+* Within the realm of typical developer knowledge
+
+.cons
+* Adds complexity to the set up of cnd
+* Negatively impacts usability, browser security pop up because of untrusted certificate
+
+According to let's encrypt project this can be partially mitigated if we serve `comit-i` on an IP address instead of on `localhost`
+Reference: https://letsencrypt.org/docs/certificates-for-localhost/
+
 
 == Implementing TLS
 
+There are two aspects to encrypting traffic between cnd to the client.
+1. The HTTP API exposed by cnd needs to be encrypted using TLS.
+2. *Our* client implementation, `comit-i` needs to be served by cnd using TLS.
+
 === cnd
 
 Adding TLS to cnd should be straightforward.
-TLS is supported by Warp (the web server framework we use).
-The only real consideration is the TLS keypair (key and certificate) and selfsigning the certificate.
-We can provide a default keypair so that cnd works out of the box with TLS.
 
-.This can be done in a few ways:
-* Put the keys in the repository - bad, private keys should be private.
-* Write a standalone tool to generate a new key.
-* Add code to cnd to generate a new keypair if one is not found i.e. in the default place or via a config option `--key=/path/to/key`and `--cert=/path/to/cert`.
+There are two parts to this:
+1. Exposing the HTTP API vial TLS (so clients can connect to cnd using TLS)
+2. Serving `comit-i` via TLS (so the browser can use `https://` addresses)
+
+Since both of these are done by calls to Warp they can both be done in the same manner.
+TLS is supported by Warp (the web server framework we use).
+Requires adding a field in the config file and command line options e.g., `--tls-key=/path/to/key` and `--tls-cert=/path/to/cert`.
+Adds an extra initial setup step for the user of cnd to ensure valid certificate is available on the host machine.
 
 === comit-i
 
-Request content for this section to be written by team member who is familiar with `comit-i` and/or React.
+`comit-i` must consume the HTTP API exposed by cnd using TLS
+I was unable to find any information on the internet on how to do this.
 
 == Recommendation
 
+The cnd stuff is straightforward.
+
 .Recommendations are as follows:
 * Implement TLS for cnd using Warp's TLS support.
 ** Enable TLS by default, add option `--no-tls` to disable it.
 ** Look for keypair in well known location, configurable via configuration file and command line options.
+* Leave generation of valid certificate up to the user of cnd but add documentation/hints on how to do this and what options are available (based on information in this spike).
+
+As for `comit-i` my findings were less positive.
+Recommend splitting `comit-i` up into separate parts for the front end and back end and re-writing the back end in Rust.
```

Comment:
Is your idea that we can then serve `comit-i` on `http` again?
I am afraid, this does not really help us :(
One of the reasons why it would be useful to serve `comit-i` via HTTPS is to make integrations with a Ledger Nano S possible (the browser only allows you to talk to a USB device on HTTPS)

Reply:
The idea is to serve `comit-i` using HTTPS.